misc/multistream-select/src/dialer_select.rs

```diff
@@ -268,6 +450,182 @@ where
     }
 }
 
+struct SimOpenPhase<R> {
+    state: SimOpenState<R>,
+}
+
+enum SimOpenState<R> {
+    SendNonce { io: MessageIO<R> },
+    FlushNonce { io: MessageIO<R>, local_nonce: u64 },
+    ReadNonce { io: MessageIO<R>, local_nonce: u64 },
+    SendRole { io: MessageIO<R>, local_role: Role },
+    FlushRole { io: MessageIO<R>, local_role: Role },
+    ReadRole { io: MessageIO<R>, local_role: Role },
+    Done,
+}
+
+/// Role of the local node after protocol negotiation.
+///
+/// Always equals [`Role::Initiator`] unless [`Version::V1SimultaneousOpen`] is
+/// used in which case node may end up in either role after negotiation.
+///
+/// See [`Version::V1SimultaneousOpen`] for details.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Role {
+    Initiator,
+    Responder,
+}
+
+impl<R> Future for SimOpenPhase<R>
+where
+    // The Unpin bound here is required because we produce a `Negotiated<R>` as the output.
+    // It also makes the implementation considerably easier to write.
+    R: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = Result<(MessageIO<R>, Role), NegotiationError>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        loop {
+            match mem::replace(&mut self.state, SimOpenState::Done) {
+                SimOpenState::SendNonce { mut io } => {
+                    match Pin::new(&mut io).poll_ready(cx)? {
+                        Poll::Ready(()) => {}
+                        Poll::Pending => {
+                            self.state = SimOpenState::SendNonce { io };
+                            return Poll::Pending;
+                        }
+                    }
+
+                    let local_nonce = rand::random();
+                    let msg = Message::Select(local_nonce);
+                    if let Err(err) = Pin::new(&mut io).start_send(msg) {
+                        return Poll::Ready(Err(From::from(err)));
+                    }
+
+                    self.state = SimOpenState::FlushNonce { io, local_nonce };
+                }
+                SimOpenState::FlushNonce {
+                    mut io,
+                    local_nonce,
+                } => match Pin::new(&mut io).poll_flush(cx)? {
```

Comment:
Do you know about the [`_unpin` APIs](https://docs.rs/futures/0.3.18/futures/sink/trait.SinkExt.html#method.poll_flush_unpin) in `SinkExt`?

Maybe we can use them here to avoid a bit of boilerplate.